src/EditorFeatures/CSharpTest/ConvertAnonymousTypeToClass/ConvertAnonymousTypeToClassTests.cs

```diff
@@ -0,0 +1,931 @@
+        public async Task CapturedTypeParameters()
+        {
+            var text = @"
+class Test<X> where X : struct
+{
+    void Method<Y>(List<X> x, Y[] y) where Y : class, new()
+    {
+        var t1 = [||]new { a = x, b = y };
+    }
+}
+";
+            var expected = @"
+class Test<X> where X : struct
+{
+    void Method<Y>(List<X> x, Y[] y) where Y : class, new()
+    {
+        var t1 = new {|Rename:NewClass|}(x, y);
+    }
+}
+
+internal class NewClass<X, Y>
+    where X : struct
+    where Y : class, new()
+{
+    public List<X> A { get; }
+    public Y[] B { get; }
+
+    public NewClass(List<X> a, Y[] b)
+    {
+        A = a;
+        B = b;
+    }
+
+    public override bool Equals(object obj)
+    {
+        var other = obj as NewClass<X, Y>;
+        return other != null &&
+               System.Collections.Generic.EqualityComparer<List<X>>.Default.Equals(A, other.A) &&
+               System.Collections.Generic.EqualityComparer<Y[]>.Default.Equals(B, other.B);
+    }
+
+    public override int GetHashCode()
+    {
+        var hashCode = -1817952719;
+        hashCode = hashCode * -1521134295 + System.Collections.Generic.EqualityComparer<List<X>>.Default.GetHashCode(A);
+        hashCode = hashCode * -1521134295 + System.Collections.Generic.EqualityComparer<Y[]>.Default.GetHashCode(B);
+        return hashCode;
+    }
+}";
+            await TestInRegularAndScriptAsync(text, expected);
+        }
+
+        [Fact, Trait(Traits.Feature, Traits.Features.CodeActionsConvertAnonymousTypeToClass)]
+        public async Task NewTypeNameCollision()
+        {
+            var text = @"
+class Test
+{
+    void Method()
+    {
+        var t1 = [||]new { a = 1, b = 2 };
+    }
+}
+
+class NewClass
+{
+}
+";
+            var expected = @"
+class Test
+{
+    void Method()
+    {
+        var t1 = new {|Rename:NewClass1|}(1, 2);
+    }
+}
+
+class NewClass
+{
+}
+
+internal class NewClass1
+{
+    public int A { get; }
+    public int B { get; }
+
+    public NewClass1(int a, int b)
+    {
+        A = a;
+        B = b;
+    }
+
+    public override bool Equals(object obj)
+    {
+        var other = obj as NewClass1;
+        return other != null &&
+               A == other.A &&
+               B == other.B;
+    }
+
+    public override int GetHashCode()
+    {
+        var hashCode = -1817952719;
+        hashCode = hashCode * -1521134295 + A.GetHashCode();
+        hashCode = hashCode * -1521134295 + B.GetHashCode();
+        return hashCode;
+    }
+}";
+            await TestInRegularAndScriptAsync(text, expected);
+        }
+
+        [Fact, Trait(Traits.Feature, Traits.Features.CodeActionsConvertAnonymousTypeToClass)]
+        public async Task TestDuplicatedName()
+        {
+            var text = @"
+class Test
+{
+    void Method()
+    {
+        var t1 = [||]new { a = 1, a = 2 };
+    }
+}
+";
+            var expected = @"
+internal class NewClass
+{
+    public int A { get; }
+    public int $1 { get; }
```

Comment:
>$1 [](start = 15, length = 2)

I didn't expect this... Is this right (this code doesn't parse)?

Reply:
> I didn't expect this... 

Me neither :)

> Is this right

That is a good question :D

Reply:
Can you trace where this `$1` comes from? I think the generated code should at least be syntactically valid. If we can make some other identifier, that'd solve the problem.

---
In reply to: [198321284](https://github.com/dotnet/roslyn/pull/28066#discussion_r198321284) [](ancestors = 198321284)

Reply:
well, it comes from teh compiler :)    I will add a check for this and will synthesize at least some sort of reasonable name :)

Reply:
Done.